### PR TITLE
Fix mask plotting when number of objects equal to number of images in a batch

### LINKED ISSUE #18832: Segmentation mask error in training batch when using multiprocess
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

I've noticed a very strange bug that occurs when starting YOLOv8 training in a seperate process (starting each trainig in a seperate multiprocess):

The generated training batches (for some reason always batch 2 and 4) generate weird training masks. I'm not sure if they are applied during the training or just appear on the training batch. It happens throughout all YOLOv8 versions.
At first I thought it has a connection with the albumentations library, but I was wrong.

I've attached the generated training batches as files.

kfold2_train_batch1_wrong:
has almost ALL masks wrong except images 0755

[Click to view image (Warning: Sensitive Content)](https://github.com/user-attachments/assets/e4b81136-f607-43e0-be52-77ebcbb331ad)

as a comparison I've uploaded kfold kfold4_train_batch_0, there you can clearly see image 0938 (top left ish) has the correct mask compared to kfold2_train_batch1_wrong.

[Click to view image (Warning: Sensitive Content)](https://github.com/user-attachments/assets/b5a8c3a3-004b-42f9-bc4c-d9d9b31df487)

the other training batch with same problem is kfold3_train_batch2, here only 2 images appear to be wrong (0691 and 0660)

[Click to view image (Warning: Sensitive Content)](https://github.com/user-attachments/assets/bd4bfcf5-05f8-4ffc-80b4-37bee938174b)

I can guarantee the mask conversion from the original dataset to YOLOv8 annotation format is correct and this isn't the problem. This only happens when using the multiprocess.

Any alternative way that I can use something which acts like multiprocess? I really need to start the trainig in a seperate process and get the results then close it. 

### Environment

```
Ultralytics 8.3.65 🚀 Python-3.11.8 torch-2.5.1+cu124 CUDA:0 (Tesla V100-SXM2-16GB, 16144MiB)
Setup complete ✅ (80 CPUs, 503.8 GB RAM, 379.5/439.0 GB disk)

OS                  Linux-5.15.0-1065-nvidia-x86_64-with-glibc2.35
Environment         Linux
Python              3.11.8
Install             pip
RAM                 503.76 GB
Disk                379.5/439.0 GB
CPU                 Intel Xeon E5-2698 v4 2.20GHz
CPU count           80
GPU                 Tesla V100-SXM2-16GB, 16144MiB
GPU count           1
CUDA                12.4

numpy               ✅ 1.26.4>=1.23.0
numpy               ✅ 1.26.4<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.10.0>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.1>=1.4.1
torch               ✅ 2.5.1>=1.8.0
torch               ✅ 2.5.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.1>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 6.1.1
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
```

### Minimal Reproducible Example

```
from ultralytics import YOLO
import torch.multiprocessing as mp
from multiprocessing import Queue

ROOT_DIR = os.getcwd()
DATA_DIR = os.path.join(ROOT_DIR,'data')
TRAINING_DIR = os.path.join(ROOT_DIR,'training')

def init_yaml(kfold_number=1):
        
    data = {
        'path': f'{DATA_DIR}/{kfold_number}',
        'train': 'train/images',
        'val': 'val/images',
    
        'names': {
            0: 'wound'
        }
    }

    with open(os.path.join(ROOT_DIR, "config.yaml"), "w") as f:
        yaml.dump(data, f, default_flow_style=False)

def train_model_function(i, result_queue):
    model = YOLO('yolov8x-seg.pt')
        
    results_training = model.train(data=os.path.join(ROOT_DIR, "config.yaml"), project=TRAINING_DIR, name="train_test_" + str(i),
                                   epochs=1, imgsz=512, batch=16, deterministic=True, plots=True, seed=1401,
                                   close_mosaic=0, augment=False, hsv_h=0, hsv_s=0, hsv_v=0, degrees=0, translate=0, scale=0, 
                                   shear=0.0, perspective=0, flipud=0, fliplr=0, bgr=0, mosaic=0, mixup=0, copy_paste=0, erasing=0,crop_fraction=0)
    r = {
        'train_map': results_training.seg.map,
        'train_map50': results_training.seg.map50,
        'train_map75': results_training.seg.map75,
        'train_precision': float(results_training.seg.p[0]),
        'train_recall': float(results_training.seg.r[0]),
        'train_f1': float(results_training.seg.f1[0]),
    }
    result_queue.put((r))

def start_training_in_process(i):
    
    result_queue = Queue()
    train_process = mp.Process(
        target=train_model_function,
        args=(i, result_queue)
    )
    train_process.name = "TrainingProcess"
    train_process.start()
    train_process.join()

    if not result_queue.empty():
        r = result_queue.get()
        return r
    else:
        print("No results returned from process")
        return None

if __name__ == '__main__':

   #loop through different kfolds of dataset
    for i in range(1,6):
        init_yaml(i)
        start_training_in_process(i)
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### LINKED ISSUE #22000: plot_images
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

# plotting.plot_images
我发现train_batch图像的分割标签出现错位现象：矩形框标签标注正确，但分割标签不对应矩形框，跑到下一个图像去了，然后发现plot_images函数中if idx.shape[0] == masks.shape[0]:  # overlap_masks=False，该条件运行时判断为True，而我的overlap_masks设置为默认值True，因为batch_idx为0、1、3、4、5、6、6、7，数量同样为8，但不满足0-8的顺序排序，从而出现错误判断

### Environment

none

### Minimal Reproducible Example

# Plot masks
                if len(masks):
                    if idx.shape[0] == masks.shape[0]:  # overlap_masks=False
                        image_masks = masks[idx]
                    else:  # overlap_masks=True
                        image_masks = masks[[i]]  # (1, 640, 640)
                        nl = idx.sum()
                        index = np.arange(nl).reshape((nl, 1, 1)) + 1
                        image_masks = np.repeat(image_masks, nl, axis=0)
                        image_masks = np.where(image_masks == index, 1.0, 0.0)

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes #22000 
Fixes #18832 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Fixes and optimizes mask handling in plot_images to correctly visualize overlapping masks and improve performance ⚡🖼️

### 📊 Key Changes
- Adjusts mask selection logic to treat masks as binary only when `masks.max() <= 1`, preventing misinterpretation of instance-labeled masks.
- Refactors overlap-mask processing to use NumPy broadcasting instead of `np.repeat`, creating per-instance binary masks more efficiently.
- Simplifies index generation with `np.arange(1, nl + 1)` for clearer and correct instance ID matching.
- Ensures output masks are explicitly `float32` for consistent downstream plotting.

### 🎯 Purpose & Impact
- Correctness: Fixes cases where instance-labeled masks (values > 1) were mishandled, leading to incorrect visualizations ✅
- Performance: Reduces memory usage and speeds up mask processing by removing redundant array repetition 🚀
- Reliability: More robust handling of overlapping masks across different mask formats, improving plotting consistency for users 📈

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help speed things along—an Ultralytics engineer will also review and assist soon.

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Since this PR appears to address a plotting bug in `ultralytics/utils/plotting.py`, please also provide a Minimum Reproducible Example (MRE) if not already included:
- A small snippet or script that reproduces the incorrect behavior on current `main`, ideally with a tiny synthetic batch where:
  - the number of masks equals the number of images, and
  - overlapped masks contain integer labels > 1
- Before/after screenshots of the plotted outputs demonstrating the fix
- Any specific shapes/dtypes for `masks` and indices used, to make reproduction straightforward

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🎯🖼️🚀

**Review by Laughing-q — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
🔥 Fantastic merge! Huge thanks to @Y-T-G for leading and @Laughing-q for the thoughtful contributions in PR [Fix and optimize mask handling in plot_images (PR #22031)](https://github.com/ultralytics/ultralytics/pull/22031).

“Simplicity is the ultimate sophistication.” — Leonardo da Vinci  
Your refactor embodies this: smarter mask selection, clean broadcasting over `np.repeat`, clear indexing, and consistent `float32` outputs. The result is sharper, correct visualizations for overlapping masks and a meaningful performance boost—benefiting users across notebooks and Ultralytics HUB.

We appreciate the care and craftsmanship. This is the kind of improvement that lifts the whole ecosystem. Thank you! 🙌
